src/test/java/org/mabrarov/exceptionsafety/FactoryMethodTest.java

```diff
@@ -41,9 +42,38 @@
 
   private OutputStream createConfiguredResource()
       throws IOException, TestResourceConfigurationException {
-    final OutputStream resource = createResource();
-    configureResource(resource);
-    return resource;
+    OutputStream resource = null;
+    Throwable throwable = null;
+    Method method = new Object() {}.getClass().getEnclosingMethod();
+    try {
+      resource = createResource();
+      configureResource(resource);
+      return resource;
+
+    } catch (final Throwable e) {
+      throwable = e;
+      Class[] classes = method.getExceptionTypes();
+
+      for (Class exceptionClass : classes) {
+        if (e.getClass().isAssignableFrom(exceptionClass)) {
+          throw e;
+        }
+      }
+      if (RuntimeException.class.isAssignableFrom(e.getClass())) {
+        throw e;
+      }
+
+      throw new AssertionError("Should never come here", e);
```

Comment:
What about `java.lang.Error` and its subclasses (like `java.lang.OutOfMemoryError`)?